File: macos/Services/MCP/Tools/Tier1/GetSampleRowsTool.swift
// GetSampleRowsTool.swift
// Gridex
//
// MCP Tool: Get sample rows from a table.

import Foundation

struct GetSampleRowsTool: MCPTool {
    let name = "get_sample_rows"
    let description = "Get sample rows from a table to help understand data shape. Default limit 10, max 100."
    let tier = MCPPermissionTier.schema

    let inputSchema: [String: Any] = [
        "type": "object",
        "properties": [
            "connection_id": [
                "type": "string",
                "description": "Connection identifier"
            ],
            "table_name": [
                "type": "string",
                "description": "Name of the table"
            ],
            "limit": [
                "type": "integer",
                "description": "Number of rows to return (default 10, max 100)",
                "default": 10,
                "minimum": 1,
                "maximum": 100
            ]
        ],
        "required": ["connection_id", "table_name"]
    ]

    func execute(params: JSONValue, context: MCPToolContext) async throws -> MCPToolResult {
        let connectionId = try extractConnectionId(from: params)

        guard let tableName = params["table_name"]?.stringValue else {
            throw MCPToolError.invalidParameters("table_name is required")
        }

        let limit = min(100, max(1, params["limit"]?.intValue ?? 10))

        let permission = await context.checkPermission(tier: tier, connectionId: connectionId)
        if let error = permission.errorMessage {
            throw MCPToolError.permissionDenied(error)
        }

        let (adapter, _) = try await context.getAdapter(for: connectionId)

        let result = try await adapter.fetchRows(
            table: tableName,
            schema: nil,
            columns: nil,
            where: nil,
            orderBy: nil,
            limit: limit,
            offset: 0
        )

        if result.rows.isEmpty {
            return MCPToolResult(text: "Table '\(tableName)' is empty.")
        }

        // Format as JSON array of objects
        var rows: [[String: String]] = []
        for row in result.rows {
            var rowDict: [String: String] = [:]
            for (idx, col) in result.columns.enumerated() {
                if idx < row.count {
                    rowDict[col.name] = row[idx].displayString
                }
            }
            rows.append(rowDict)
        }

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(rows)
        let json = String(data: data, encoding: .utf8) ?? "[]"

        let header = "Sample \(rows.count) row(s) from '\(tableName)':\n"
        let columns = "Columns: " + result.columns.map { "\($0.name) (\($0.dataType))" }.joined(separator: ", ") + "\n\n"

        return MCPToolResult(text: header + columns + json)
    }
}

File: windows/Gridex/Services/MCP/Tools/Tier1/GetSampleRowsTool.cpp
//
// GetSampleRowsTool.cpp
//

#include "GetSampleRowsTool.h"
#include "../MCPToolHelpers.h"
#include <algorithm>

namespace DBModels
{
    MCPToolResult GetSampleRowsTool::execute(const nlohmann::json& params, MCPToolContext& ctx)
    {
        const auto connId = MCPTool::extractConnectionId(params);

        if (!params.contains("table_name") || !params["table_name"].is_string())
            throw MCPToolError::invalidParameters("table_name is required");
        const auto tableUtf8 = params["table_name"].get<std::string>();
        const std::wstring tableW = MCPToolHelpers::fromUtf8(tableUtf8);

        const std::wstring schemaW = (params.contains("schema") && params["schema"].is_string())
            ? MCPToolHelpers::fromUtf8(params["schema"].get<std::string>()) : std::wstring{};

        int limit = 10;
        if (params.contains("limit") && params["limit"].is_number_integer())
            limit = (std::min)(100, (std::max)(1, params["limit"].get<int>()));

        const auto perm = ctx.checkPermission(tier(), connId);
        if (const auto* msg = perm.errorMessage())
            throw MCPToolError::permissionDenied(*msg);

        auto [adapter, config] = ctx.getAdapter(connId);

        // fetchRows signature: (table, schema, limit, offset, orderBy, ascending).
        QueryResult r = adapter->fetchRows(tableW, schemaW, limit, 0, L"", true);
        if (!r.success)
            throw MCPToolError::queryFailed(MCPToolHelpers::toUtf8(r.error));

        if (r.rows.empty())
            return MCPToolResult::text("Table '" + tableUtf8 + "' is empty.");

        // Columns header summary.
        std::string cols = "Columns: ";
        for (size_t i = 0; i < r.columnNames.size(); ++i)
        {
            if (i) cols += ", ";
            cols += MCPToolHelpers::toUtf8(r.columnNames[i]);
            if (i < r.columnTypes.size())
                cols += " (" + MCPToolHelpers::toUtf8(r.columnTypes[i]) + ")";
        }

        nlohmann::json rows = nlohmann::json::array();
        for (const auto& row : r.rows)
        {
            nlohmann::json o = nlohmann::json::object();
            for (const auto& cname : r.columnNames)
            {
                auto it = row.find(cname);
                const std::wstring v = (it != row.end()) ? it->second : L"";
                o[MCPToolHelpers::toUtf8(cname)] =
                    isNullCell(v) ? nlohmann::json(nullptr)
                                  : nlohmann::json(MCPToolHelpers::toUtf8(v));
            }
            rows.push_back(std::move(o));
        }

        return MCPToolResult::text(
            "Sample " + std::to_string(r.rows.size()) + " row(s) from '" +
            tableUtf8 + "':\n" + cols + "\n\n" + rows.dump(2));
    }
}
